crates/swc_common/src/plugin.rs

```diff
@@ -8,38 +8,58 @@ use crate::{syntax_pos::Mark, SyntaxContext};
 use anyhow::Error;
 use std::any::type_name;
 
-/// Serialize ast to pass into plugins. This is being called from host side.
-/// TODO: https://github.com/swc-project/swc/issues/3167
-#[cfg(feature = "plugin-base")]
-pub fn serialize_for_plugin<T>(t: &T) -> Result<rkyv::AlignedVec, Error>
+/// A struct wraps internal representation of serialized Program to be passed
+/// across plugin's host to guest. Plugin author should not rely on specific
+/// details serialized byte format struct contains: it is strictly
+/// implementation detail which can change anytime.
+#[cfg(any(feature = "plugin-mode", feature = "plugin-rt"))]
+pub struct SerializedProgram(pub rkyv::AlignedVec);
+
+impl SerializedProgram {
+    #[cfg(any(feature = "plugin-mode", feature = "plugin-rt"))]
+    pub fn new(bytes: Vec<u8>, len: i32) -> SerializedProgram {
+        let mut vec = rkyv::AlignedVec::with_capacity(
+            len.try_into()
```

Comment:
What should happen if `bytes.len() != len`?
Also, is there a reason to have `len` when we can extract it from `bytes.len()`?

Reply:
vec's len is actually constructed from `len` which is returned from raw bytes ptr + size, so I expect length difference would not occur. I decided to accept size explicitly though, as it's easier to reason about we are constructing a struct from raw ptr instead of vec itself. Not a specific technical reason we have to but I felt it gives better implication.

Reply:
Oh, I see. Thanks!